arrow-libs/core/arrow-core-data/src/main/kotlin/arrow/core/NonEmptyList.kt

```diff
@@ -550,8 +429,188 @@ object NonEmptyListSemigroup : Semigroup<NonEmptyList<Any?>> {
+    NonEmptyList.unit,
+    NonEmptyList.unit
+  ) { aa, bb, cc, dd, ee, _, _, _, _, _ ->
+    map(aa, bb, cc, dd, ee)
+  }
+
+inline fun <A, B, C, D, E, F, Z> NonEmptyList<A>.zip(
+  b: NonEmptyList<B>,
+  c: NonEmptyList<C>,
+  d: NonEmptyList<D>,
+  e: NonEmptyList<E>,
+  f: NonEmptyList<F>,
+  map: (A, B, C, D, E, F) -> Z
+): NonEmptyList<Z> =
+  zip(
+    b, c, d, e, f,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit
+  ) { aa, bb, cc, dd, ee, ff, _, _, _, _ ->
+    map(aa, bb, cc, dd, ee, ff)
+  }
+
+inline fun <A, B, C, D, E, F, G, Z> NonEmptyList<A>.zip(
+  b: NonEmptyList<B>,
+  c: NonEmptyList<C>,
+  d: NonEmptyList<D>,
+  e: NonEmptyList<E>,
+  f: NonEmptyList<F>,
+  g: NonEmptyList<G>,
+  map: (A, B, C, D, E, F, G) -> Z
+): NonEmptyList<Z> =
+  zip(
+    b, c, d, e, f, g,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit
+  ) { aa, bb, cc, dd, ee, ff, gg, _, _, _ ->
+    map(aa, bb, cc, dd, ee, ff, gg)
+  }
+
+inline fun <A, B, C, D, E, F, G, H, Z> NonEmptyList<A>.zip(
+  b: NonEmptyList<B>,
+  c: NonEmptyList<C>,
+  d: NonEmptyList<D>,
+  e: NonEmptyList<E>,
+  f: NonEmptyList<F>,
+  g: NonEmptyList<G>,
+  h: NonEmptyList<H>,
+  map: (A, B, C, D, E, F, G, H) -> Z
+): NonEmptyList<Z> =
+  zip(
+    b, c, d, e, f, g, h,
+    NonEmptyList.unit, NonEmptyList.unit
+  ) { aa, bb, cc, dd, ee, ff, gg, hh, _, _ ->
+    map(aa, bb, cc, dd, ee, ff, gg, hh)
+  }
+
+inline fun <A, B, C, D, E, F, G, H, I, Z> NonEmptyList<A>.zip(
+  b: NonEmptyList<B>,
+  c: NonEmptyList<C>,
+  d: NonEmptyList<D>,
+  e: NonEmptyList<E>,
+  f: NonEmptyList<F>,
+  g: NonEmptyList<G>,
+  h: NonEmptyList<H>,
+  i: NonEmptyList<I>,
+  map: (A, B, C, D, E, F, G, H, I) -> Z
+): NonEmptyList<Z> =
+  zip(
+    b, c, d, e, f, g, h, i,
+    NonEmptyList.unit
+  ) { aa, bb, cc, dd, ee, ff, gg, hh, ii, _ ->
+    map(aa, bb, cc, dd, ee, ff, gg, hh, ii)
+  }
+
+inline fun <A, B, C, D, E, F, G, H, I, J, Z> NonEmptyList<A>.zip(
+  b: NonEmptyList<B>,
+  c: NonEmptyList<C>,
+  d: NonEmptyList<D>,
+  e: NonEmptyList<E>,
+  f: NonEmptyList<F>,
+  g: NonEmptyList<G>,
+  h: NonEmptyList<H>,
+  i: NonEmptyList<I>,
+  j: NonEmptyList<J>,
+  map: (A, B, C, D, E, F, G, H, I, J) -> Z
+): NonEmptyList<Z> {
+  val buffer = ArrayList<Z>()
+  for (aa in tail) {
+    for (bb in b.tail) {
+      for (cc in c.tail) {
+        for (dd in d.tail) {
+          for (ee in e.tail) {
+            for (ff in f.tail) {
+              for (gg in g.tail) {
+                for (hh in h.tail) {
+                  for (ii in i.tail) {
+                    for (jj in j.tail) {
+                      buffer.add(map(aa, bb, cc, dd, ee, ff, gg, hh, ii, jj))
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+  return NonEmptyList(
+    map(head, b.head, c.head, d.head, e.head, f.head, g.head, h.head, i.head, j.head),
+    buffer
+  )
+}
```

Comment:
```suggestion
): NonEmptyList<Z> =
  NonEmptyList(
    map(b.head, c.head, d.head, e.head, f.head, g.head, h.head, i.head, j.head, k.head),
    b.tail.zip(c.tail, d.tail, e.tail, f.tail, e.tail, f.tail, g.tail, h.tail, i.tail, j.tail, k.tail, map)
  )
```

Reply:
I think this isn't right, you're missing the value for `this`